pkg/custom-provider/external_info_map.go

```diff
@@ -0,0 +1,128 @@
+	MetricName() string
+	//WithSeries associates the provided labels with this metric.
+	WithSeries(labels labels.Set)
+	//WithNamespacedSeries associates the provided labels with this metric, but within a particular namespace.
+	WithNamespacedSeries(namespace string, labels labels.Set)
+	//Exports a collection of all the metrics currently being tracked.
+	ExportMetrics() []ExportedMetric
+	//Generates a query to select the series/values for the metric this object represents.
+	GenerateQuery(namespace string, selector labels.Selector) (prom.Selector, error)
+}
+
+type externalInfoMap struct {
+	metrics map[string]ExternalMetricData
+}
+
+type externalMetricData struct {
+	metricName     string
+	namespacedData map[string][]labels.Set
+	generatedBy    MetricNamer
+}
+
+//NewExternalMetricData creates an ExternalMetricData for the provided metric name and namer.
+func NewExternalMetricData(metricName string, generatedBy MetricNamer) ExternalMetricData {
+	return &externalMetricData{
+		metricName:     metricName,
+		generatedBy:    generatedBy,
+		namespacedData: map[string][]labels.Set{},
+	}
+}
+
+//NewExternalInfoMap creates an empty ExternalInfoMap for storing external metric information.
+func NewExternalInfoMap() ExternalInfoMap {
+	return &externalInfoMap{
+		metrics: map[string]ExternalMetricData{},
+	}
+}
+
+func (i *externalInfoMap) ExportMetrics() []ExportedMetric {
+	results := make([]ExportedMetric, 0)
+	for _, info := range i.metrics {
+		exported := info.ExportMetrics()
+		results = append(results, exported...)
+	}
+
+	return results
+}
+
+func (i *externalInfoMap) FindMetric(metricName string) (data ExternalMetricData, found bool) {
+	data, found = i.metrics[metricName]
+	return data, found
+}
+
+func (i *externalInfoMap) TrackMetric(metricName string, generatedBy MetricNamer) ExternalMetricData {
+	data, found := i.metrics[metricName]
+	if !found {
+		data = NewExternalMetricData(metricName, generatedBy)
+		i.metrics[metricName] = data
+	}
+
+	return data
+}
+
+func (d *externalMetricData) MetricName() string {
+	return d.metricName
+}
+
+func (d *externalMetricData) GenerateQuery(namespace string, selector labels.Selector) (prom.Selector, error) {
+	return d.generatedBy.QueryForExternalSeries(namespace, d.metricName, selector)
+}
+
+func (d *externalMetricData) ExportMetrics() []ExportedMetric {
+	results := make([]ExportedMetric, 0)
+	for namespace, labelSets := range d.namespacedData {
+		for _, labelSet := range labelSets {
+			results = append(results, ExportedMetric{
+				Labels:     labelSet,
+				MetricName: d.metricName,
+				Namespace:  namespace,
+			})
+		}
+	}
+
+	return results
+}
+
+func (d *externalMetricData) WithSeries(labels labels.Set) {
+	d.WithNamespacedSeries("", labels)
+}
+
+func (d *externalMetricData) WithNamespacedSeries(namespace string, seriesLabels labels.Set) {
+	data, found := d.namespacedData[namespace]
+	if !found {
+		data = []labels.Set{}
+	}
+
+	data = append(data, seriesLabels)
```

Comment:
You should be able to determine those label names and values from the Prometheus response -- Prometheus returns something like `{"value": 1.23, "labels": {"foo": "bar"}}`, so you should be able to use the returned `labels` field to populate label values.

Reply:
The discovery mechanism within `k8s-prometheus-adapter` (the one driven by the config) doesn't _need_ the label names and values to operate. You're definitely correct with respect to the "live results", where we make a request out to Prometheus to get metric values. We can just pluck the labels off of the result. So I'm with you up to that point, at least.

The thing where I get lost/confused is this:
I followed in the footsteps of `basicSeriesRegistry`, in that when the series are discovered/named/etc, the results are cached. As such, when the request comes in to `ListAllExternalMetrics() []ExternalMetricInfo`, it's not currently making a request out to Prometheus to find series (in part because we'd have to re-transform them to conform to the configuration), but instead it's returning the previously cached series information from the last update cycle.

If we throw away the label names/values from the _cache_ then we won't have them to include in the response from the external metrics API (for that endpoint, in particular).

Are you suggesting that we toss the labels/values and _don't_ return them from the `ListAllExternalMetrics() []ExternalMetricInfo` endpoint?

Reply:
The person who implemented `ListAllExternalMetrics` put extra unnecessary info in there, and I missed it during review.  For now, just don't populate the label set, and we'll go and fix it in custom-metrics-apiserver

Reply:
Ah, it all makes sense now.

Perfect. I'll get right on that.